composer

```diff
@@ -489,16 +501,39 @@ if (!class_exists('ComposerWrapper')) {
             \fwrite(STDERR, $text . "\n");
         }
 
+        protected function isSelfUpdate($cliArguments)
+        {
+            foreach ($cliArguments as $cliArgument) {
+                if ('-' === \substr($cliArgument, 0, 1)) {
+                    continue;
+                }
+
+                // If the first argument that is not a CLI option (i.e. not starts with dash, '-'),
+                // and it's self-update, it should be run in a subprocess rather than delegated
+                // because otherwise wrapper will be replaced
+                if ('self-update' === $cliArgument || 'selfupdate' === $cliArgument) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
         /**
          * @throws Exception
          */
-        public function run()
+        public function run($cliArguments)
         {
             $this->params->loadReal();
             $composerPathName = "{$this->params->getComposerDir()}/composer.phar";
 
             $this->ensureInstalled($composerPathName);
             $this->ensureExecutable($composerPathName);
+
+            if ($this->isSelfUpdate($cliArguments)) {
+                return $this->runByRealComposerSubprocess($composerPathName, $cliArguments);
+            }
+
```

Comment:
Do you think the `delegate` (php's require)  call below has some significant benefits in compare with `runByRealComposerSubprocess` (passthru)? 

I think it could be good idea always use passthru to prevent future similar errors. And remove some existed hacks (https://github.com/kamazee/composer-wrapper/blob/master/composer#L463) 
Only one think that could be broken, it is cwd for wrapper and composer may be different. 
So, probably something like this helps:
```
$cwd = getcwd();
passthru("cd $cwd && {OTHER_COMPOSER_CALL_LINE}"); 
```
ps Actually, up to this PR I thought that `passthru` is always used 🙉

Reply:
> Do you think the delegate (php's require) call below has some significant benefits in compare with runByRealComposerSubprocess (passthru)?

If a php interpreter is called with specific options (e.g. `php -d memory_limit=2G ./composer update`: increasing available memory is rather common for Composer 1), it's not possible to replicate them easily for a subprocess